librocksdb_sys/crocksdb/c.cc

```diff
@@ -2188,6 +2195,103 @@ DB::RunManualCompaction
 custom cache
 table_properties_collectors
 */
+struct crocksdb_tablepropertiescollectorfactorycontext_t {
+    TablePropertiesCollectorFactory::Context rep;
+};
+
+struct crocksdb_tablepropertiescollector_t : public TablePropertiesCollector {
+    void *state_;
+    void (*destructor_)(void *);
+    void (*add_userkey_)(
+            void *,
+            const char *key, size_t key_length,
+            const char *value, size_t value_length,
+            int entry_type, uint64_t seq,
+            uint64_t file_size);
+    void (*finish_)(void *, char **keys, int *key_count, char **value, int *value_count);
+    void (*readable_properties_)(void *);
+    unsigned char need_compact_;
+    const char *(*name_)(void *);
+
+    virtual ~crocksdb_tablepropertiescollector_t() {
+        (*destructor_)(state_);
+    }
+
+    virtual Status AddUserKey(const Slice &key, const Slice &value,
+                              EntryType type, SequenceNumber seq,
+                              uint64_t file_size) override {
+        // For backwards-compatibility.
+        (*add_userkey_)(
+                state_,
+                key.data(), key.size(),
+                value.data(), value.size(),
+                type,
+                seq,
+                file_size);
+        return Status::OK();
+    }
+
+    virtual Status Finish(UserCollectedProperties *properties) override {
+        char **keys, **values;
+        int keys_count, value_count;
+        (*finish_)(state_, keys, &keys_count, values, &value_count);
+        if (keys == NULL || values == NULL) {
+            return Status::OK();
+        }
+        assert(keys_count == value_count);
+        for (int i = 0; i < keys_count; i++) {
+            properties->insert(std::make_pair(std::string(keys[i]), std::string(values[i])));
+            free(keys[i]);
+            free(values[i]);
+        }
+        // free the memory
+        free(keys);
```

Comment:
To avoid creating CString(which may be expensive), add const size_t* keys_lengths as extra parameter.